src/control/common/pci_utils.go

```diff
@@ -37,3 +47,302 @@ func ParsePCIAddress(addr string) (dom, bus, dev, fun uint64, err error) {
 
 	return
 }
+
+// PCIAddress represents a standard PCI address with domain and BDF.
+type PCIAddress struct {
+	Domain   string
+	Bus      string
+	Device   string
+	Function string
+}
+
+func (pa *PCIAddress) String() string {
+	if pa == nil {
+		return ""
+	}
+
+	return fmt.Sprintf("%s:%s:%s.%s", pa.Domain, pa.Bus, pa.Device, pa.Function)
+}
+
+// Equals compares string representation of address.
+func (pa *PCIAddress) Equals(o *PCIAddress) bool {
+	if pa == nil || o == nil {
+		return false
+	}
+
+	return o.String() == pa.String()
+}
+
+// IsVMDBackingAddress indicates whether PCI address is a VMD backing device.
+func (pa *PCIAddress) IsVMDBackingAddress() bool {
+	if pa == nil {
+		return false
+	}
+
+	return pa.Domain != "0000"
+}
+
+// BackingToVMDAddress returns the VMD PCI address associated with a VMD backing devices address.
+func (pa *PCIAddress) BackingToVMDAddress() (*PCIAddress, error) {
+	if pa == nil {
+		return nil, errors.New("PCIAddress is nil")
+	}
+	if !pa.IsVMDBackingAddress() {
+		return nil, errors.New("not a vmd backing device address")
+
+	}
+
+	// assume non-zero pci address domain field indicates a vmd backing device wisth
+	// fixed length field
+	if len(pa.Domain) != vmdDomainLen {
+		return nil, errors.New("unexpected length of vmd domain")
+	}
+
+	return NewPCIAddress(fmt.Sprintf("0000:%s:%s.%s", pa.Domain[:2], pa.Domain[2:4],
+		pa.Domain[5:]))
+}
+
+// LessThan evaluate whether "this" address is less than "other" by comparing
+// domain/bus/device/function in order.
+func (pa *PCIAddress) LessThan(other *PCIAddress) bool {
+	if pa == nil || other == nil {
+		return false
+	}
+
+	if pa.Domain != other.Domain {
+		return hexStr2Int(pa.Domain) < hexStr2Int(other.Domain)
+	}
+
+	if pa.Bus != other.Bus {
+		return hexStr2Int(pa.Bus) < hexStr2Int(other.Bus)
+	}
+
+	if pa.Device != other.Device {
+		return hexStr2Int(pa.Device) < hexStr2Int(other.Device)
+	}
+
+	return hexStr2Int(pa.Function) < hexStr2Int(other.Function)
+}
+
+// NewPCIAddress creates a PCIAddress struct from input string.
+func NewPCIAddress(addr string) (*PCIAddress, error) {
+	dom, bus, dev, fun, err := parsePCIAddress(addr)
+	if err != nil {
+		return nil, err
```

Comment:
Minor nit: From a usability standpoint, it would be better to include the entire address in the error message, given that the constructors accept lists of addresses. So instead of returning an error like `strconv.ParseUint: parsing "gg": invalid syntax`, you could wrap the parse error with more information, e.g. `unable to parse "0000:gg:00.0": strconv.ParseUint: parsing "gg": invalid syntax`.

This would allow the user to quickly identify the bad address(es) in the list.